# Data Exploration

Import các thư viện cần thiết

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

Đọc dữ liệu từ file `Data/2019.csv`

In [2]:
DATA_PATH = 'Data/Superstore.csv'

data = pd.read_csv(DATA_PATH, encoding='ISO-8859-1')
print("Shape: ", data.shape)
data.head()
data.info()

Shape:  (9994, 21)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   fl

➡️ **Nhận xét**: 

- Dữ liệu có 9994 dòng và 21 cột.

- Mỗi dòng tượng trưng cho một hàng hóa cụ thể trong một đơn hàng (mỗi đơn hàng có thể gồm nhiều dòng) và các thông tin liên quan là `Order ID` (mã đơn hàng) `Customer Name`(Tên khách hàng), `Product Name`(Tên sản phẩm), ...

- Kích thước dữ liệu: 2.18 MB.

Số giá trị duy nhất của cột `Row ID`:

In [3]:
print("Number of unique values in column 'Row ID' ", data['Row ID'].nunique())

Number of unique values in column 'Row ID'  9994


➡️ **Nhận xét**: 

- Số giá trị duy nhất của cột `Row ID` bằng với số dòng của dataset (9994) → không có dòng dữ liệu trùng.

Dữ liệu có các cột nào và ý nghĩa từng cột là gì?

In [4]:
data.columns.tolist()

['Row ID',
 'Order ID',
 'Order Date',
 'Ship Date',
 'Ship Mode',
 'Customer ID',
 'Customer Name',
 'Segment',
 'Country',
 'City',
 'State',
 'Postal Code',
 'Region',
 'Product ID',
 'Category',
 'Sub-Category',
 'Product Name',
 'Sales',
 'Quantity',
 'Discount',
 'Profit']

➡️ **Nhận xét**: 

- Ý nghĩa của từng đặc trưng:

    - Dưới đây là bảng giải thích chi tiết ý nghĩa của 18 cột dữ liệu trong file train.csv. Mình đã chia nhóm các cột theo chủ đề để bạn dễ hình dung cấu trúc của bảng dữ liệu này.

1. Nhóm thông tin về Đơn hàng (Order Information)

- `Row ID` (Số thứ tự dòng): Mã định danh duy nhất cho từng dòng trong bảng dữ liệu.
    - Ý nghĩa: Cột này không có ý nghĩa phân tích, chỉ dùng để đánh số thứ tự (Index).

- `Order ID` (Mã đơn hàng): Mã số duy nhất đại diện cho một lần thanh toán (một giỏ hàng).
    - Ý nghĩa: Một Order ID có thể xuất hiện nhiều lần ở nhiều dòng khác nhau nếu đơn hàng đó có nhiều sản phẩm. Ví dụ: Đơn hàng `Order ID` = 'CA-2017-152156' có 2 dòng (mua nhiều hơn 1 sản phẩm).

- `Order Date` (Ngày đặt hàng): Ngày khách hàng chốt đơn trên hệ thống.
    - Ý nghĩa: Dùng để phân tích chuỗi thời gian (Time Series), tính doanh số theo ngày/tháng/năm.

- `Ship Date` (Ngày giao hàng): Ngày hàng hóa được giao đến người nhận.
    - Ý nghĩa: Dùng để tính toán thời gian xử lý đơn hàng từ lúc đặt đến lúc giao(`Ship Date` - `Order Date`).

- `Ship Mode` (Hình thức vận chuyển): Loại dịch vụ giao hàng mà khách đã chọn.
    - Ý nghĩa: Dùng để phân ra thành các loại như: **Standard Class** (Giao thường), **Second Class** (Giao nhanh vừa), **First Class** (Giao ưu tiên), **Same Day** (Giao trong ngày).

- `Customer ID` (Mã khách hàng): Mã định danh duy nhất cho mỗi khách hàng.
    - Ý nghĩa: Giúp phân biệt khách hàng khác nhau.

- `Customer Name` (Tên khách hàng): Tên của khách hàng.
    - Ý nghĩa: Dùng để lưu trữ thông tin tên khách hàng khi cần.

- `Segment` (Phân khúc khách hàng): Nhóm các loại khách hàng.
    - Ý nghĩa: Phân loại khách hàng thành 3 nhóm chính: **Consumer** (Người tiêu dùng), **Corporate** (Doanh nghiệp), **Home Office** (Văn phòng tại nhà). Dùng để phân tích hành vi tiêu dùng khác nhau giữa các nhóm.

- `Country` (Quốc gia): Tên quốc gia.
    - Ý nghĩa: Xác định thị trường cấp quốc gia (chủ yếu là United States).

- `City` (Thành phố): Tên thành phố.
    - Ý nghĩa: Dùng để phân tích hiệu quả kinh doanh giữa các thành phố.

- `State` (Tiểu bang): Tên tiểu bang.
    - Ý nghĩa: Dùng để so sánh hiệu suất kinh doanh giữa các bang.

`Postal Code` (Mã bưu chính): Mã số định vị khu vực giao hàng (Zip code).
    - Ý nghĩa: Dùng để định vị chính xác vị trí địa lý trong quá trình vận chuyển.

`Region` (Vùng miền): Khu vực địa lý rộng lớn.
    - Ý nghĩa: Thị trường chia thành 4 vùng lớn: South (Nam), West (Tây), Central (Trung tâm), East (Đông).

- `Product ID` (Mã sản phẩm): Mã định danh cho từng sản phẩm.
    - Ý nghĩa: Dùng để quản lý sản phẩm khác nhau.

- `Category` (Danh mục chính): Phân loại sản phẩm liên quan thành các danh mục chính cụ thể.
    - Ý nghĩa: Phân loại sản phẩm thành 3 nhóm lớn: **Furniture** (Nội thất), **Office Supplies** (Văn phòng phẩm), **Technology** (Công nghệ).

- `Sub-Category` (Danh mục phụ): Các nhánh nhỏ hơn trong danh mục chính.
    - Ý nghĩa: Cung cấp chi tiết hơn về loại sản phẩm (Ví dụ: Trong **Furniture** sẽ có Chairs, Tables, Bookcases...).

- `Product Name` (Tên sản phẩm): Tên của sản phẩm.
    - Ý nghĩa: Tên chi tiết của sản phầm khách hàng quan tâm.

- `Sales` (Doanh thu - Revenue): Tổng số tiền thu được từ dòng sản phẩm đó trong đơn hàng.
    - Ý nghĩa: Là chỉ số quan trọng nhất để đo lường quy mô tăng trưởng.

- `Quantity` (Số lượng): Số lượng sản phẩm khách đã mua.
    - Ý nghĩa: Dùng để lưu số lượng các sản phảm.

- `Discount` (Chiết khấu/Giảm giá): Phần trăm giảm giá áp dụng cho sản phẩm nào đó.
    - Ý nghĩa: Dùng để phân tích chiến lược giá và tác động của khuyến mãi lên lợi nhuận.

- `Profit` (Lợi nhuận): Số tiền lãi (hoặc lỗ) thu được từ giao dịch này.
    - Ý nghĩa: Tính được sau khi trừ đi chi phí. Nếu giá trị dương là Lãi, âm là Lỗ. Đây là chỉ số quan trọng để đánh giá hiệu quả kinh doanh thực tế (đôi khi Doanh số cao nhưng Lợi nhuận âm do giảm giá quá sâu).

Kiểu dữ liệu của các cột là gì?

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

➡️ **Nhận xét**: 

- Các cột: `Order Date`, `Ship Date` Hiện tại đang ở dạng object (chuỗi ký tự - string).
    - Cẩn được ép kiểu về **datetime** hoặc tại đặc trưng mới như là `Order_year`, `Ship_Month`,...
    - Có thể tạo đặc trưng mới là thời gian sử lý đơn hàng: `Processing_Time` = `Ship Date` - `Order Date` (để đánh giá hiệu quả giao và xử lí đơn hàng). Kiểm tra các đơn hàng có `Ship Date` < `Order Date` (lỗi logic) để loại bỏ hoặc sửa chữa.

- Các cột: `Ship Mode`, `Segment`, `Region`, `Category` chứa dữ liệu dạng phân loại.
    - Có thể sử dụng **One-hot encoding** để xử lý.

- Các cột: `City`, `State`, `Sub-Category`
    - Có thể sử dụng các chiến lược: `Frequency Encoding` hay `Target Encoding` để xử lý.

- Các cột: `Sales`, `Quantity`, `Discount`, `Profit` đang ở dữ liệu số định dạng đúng của nó.
    - Có thể kiểm tra và xử lý các giá trị ngoại lai (outliers).
    - Chuẩn hóa giá trị cho phù hợp để huấn luyện mô hình.
    - `Sales` và `Profit` thường có phân phối bị lệch (skewed). Có thể áp dụng **Log Transformation** (lấy logarit) để đưa về phân phối chuẩn giúp mô hình hồi quy hoạt động tốt hơn.

- Các cột: `Row ID`, `Order ID`, `Customer ID`, `Product ID`, `Customer Name`, `Product Name`, `Postal Code` đều mang ý nghĩa là định danh với các đối tượng cùng loại có thể không hữu dụng khi huấn luyện mô hình. 
    - **Lưu ý**: `Postal Code` đang ở dạng số nhưng ý nghĩa nó không dùng để tính toán cộng hay trừ mà chỉ để định danh, cần ép kiểu thành chuỗi.